### PR TITLE
Create UploadCsvBox Component #1941

### PR DESCRIPTION
**Pull request by bolom**
Implement a React component, UploadCsvBox, that handles CSV file uploads. The component sends the uploaded file to the backend for validation instead of parsing it in the frontend.
 It displays the filename and provides information about the number of uploaded samples. In case any batches are not found in the CSV, it highlights the relevant batches and shows the corresponding missing UUIDs in a tooltip. 

I'm facing a few CSS issues and could use some help to fix them. But before we dive into that, I'd love to hear your thoughts on the code.

### CODE REVIEW

**Comment by leandroradusky** [comment]
Hi! I was testing this feature and functionally it works as expected. Codewise for me it makes sense but I leave @ysbaddaden to make further comments on that.

Regarding the CSSs, [this](https://github.com/instedd/cdx/pull/1903/commits/c6d09b62166303cdc6c94784a87a897949af89a7) commit, where all was implemented in the frontend, had the correct CSS styling. It will be not straightforward since you will have to dig a little on the JS, but the rows showing the info about the uploaded CSVs where properly shown there, hope that is helpful :)

**Comment by bolom** [comment]
@leandroradusky improve the css, can you take a look

**Comment by leandroradusky** [comment]
Hi @bolom! I've taken a look and here are some comments (we're close :smile:)

image

The rows appear to have a max width and not go beyond that point, while the file name should be left aligned and the info in the right. When there are missing batches and long file names the message overlap.  Another place to "steal" styles & behavior is available in `samples>individual samples>upload results`. There, it looks like this:

image

Another things to mention:
- The `Add file` button is just text and the `+` button, as it looks above, no "Choose file" button should be displayed. 
- The `Add file` button is after the list of files added and not before.
- When there are zero samples it is saying `0 sample` missing the final `s`.

I think this are only things :)

**Comment by bolom** [comment]
@leandroradusky let me know what do you think. I've updated the css

**Comment by leandroradusky** [comment]
Hi @bolom! It's almost perfect!

Two minimal things:

image

- When more than one batch is not found a `<br>` is displayed instead of a new line for each batch.
- If it's not a pain (since is not that important) the `Add file` plus circle should be aligned with the minus circle of the files.

:)

File: app/controllers/csv_validations_controller.rb
[the remark this answers is not in the document]
Comment (ysbaddaden) [reply app/controllers/csv_validations_controller.rb:1]: There is also a security issue: it doesn't ensure that the batches are accessible by the user!

**Review by ysbaddaden — COMMENTED** [review]
This is much more complex that I anticipated. I merely proposed to parse the CSV quickly on the frontend (no background HTTP request), then onsubmit let the backend parse the CSV, validate its contents, and re-render the form with errors when needed.

This behavior, closer to how Stimulus works, actually fits better into the overall style of CDx.

Now, I let @leandroradusky and @devduarte be the judge.

File: app/controllers/csv_validations_controller.rb
[the remark this answers is not in the document]
Comment (bolom) [reply app/controllers/csv_validations_controller.rb:1]: > CsvValidations isn't a resource and the name can be confusing: is it validating that the CSV is just a valid CSV text file? Same for the #create action which isn't creating anything.
I should have a method on the boxes controller. For example boxes#validate_csv.

I'm disagree with you ^^
I see CsvValidations as a resource in its own right, providing a CSV validation service - a 'service resource' if you will. The create action may seem unconventional, but it's 'creating' a new "validation", so it fits within CRUD boundaries.
Maintaining CRUD methods helps us avoid 'Fat Controllers' and keeps our code tidy

[the remark this answers is not in the document]
Comment (bolom) [reply app/controllers/csv_validations_controller.rb:1]: We each have our own, and that's absolutely okay!

Whenever I feel the urge to add a method to a controller that doesn't fall into the pre-established five REST actions we typically use, my go-to move is to create a new controller. I just go ahead and give it a name that suits it.

**Comment by bolom** [comment]
@leandroradusky ready for your review

**Comment by leandroradusky** [comment]
In the last commit:

- Removed unused vars and commented code lines
- Fixed the pluralization of batches and samples
- The number of samples was displayed incorrectly, it was displaying the number of distinct batches instead.
- The saving of boxes for the CSV mode was broken when any batches were not found, fix the `batches_data` method and the validations.

While I don't dare to know which controller flavor is better and more correct rubywise, I do agree that `CsvValidation` should be renamed to `CsvBoxValidation` to accomplish a better description of what the module does. I've tried to change it without success, I would like to have these changes merged to deliver to the client, we can create an issue for the renaming after.

Also, have no idea what is this lint test failing :scream: :cry:

**Comment by ysbaddaden** [comment]
@leandroradusky About the lint errors: one can be corrected locally with `rubocop -A`, for the others, you can search for `rubocop Lint/NonLocalExitFromIterator` and you should find some explanation. For example: https://docs.rubocop.org/rubocop/cops_lint.html#lintnonlocalexitfromiterator

**Comment by bolom** [comment]
> In the last commit:
> 
>     * Removed unused vars and commented code lines
> 
>     * Fixed the pluralization of batches and samples
> 
>     * The number of samples was displayed incorrectly, it was displaying the number of distinct batches instead.
> 
>     * The `<br>` between batches was still displayed.
> 
>     * The saving of boxes for the CSV mode was broken when any batches were not found, fix the `batches_data` method and the validations.
> 
> 
> While I don't dare to do an opinion of which controller flavor is better and more correct rubywise, I do agree that `CsvValidation` should be renamed to `CsvBoxValidation` to accomplish a better description of what the module does. I've tried to change it without success, I would like to have these changes merged to deliver to the client, we can create an issue for the renaming after.
> 
> Also, have no idea what is this lint test failing 😱 😢

you can also run this command
bundle exec rubocop -A

**Comment by ysbaddaden** [comment]
Ideally the endpoint should be `POST /boxes/validate`. No need to specify CSV, it's a mere format for the data; this is similar to how we don't have `create_form` or `create_json` but merely `create`. We can also introduce support for other formats (e.g. XLSX) on the same endpoint.

We could try to make another controller, but then the action becomes too weird:

- `boxes/validate#validate` (no: pointless repetition)
- `boxes/validations#validate` (no: the validations term comes from nowhere)
- `boxes/validate#create` (nope: create means that we create a resource in CDx)

So, I still stand on my suggestion of just adding a method to the Boxes controllers. It's better to group related logic together (creation and validation are _very_ related), and it fits better into the project as it stands.

@leandroradusky we can push that to a followup issue.

**Pull request merged**